packages/techdocs-common/src/helpers.ts

```diff
@@ -50,9 +60,113 @@ export const parseReferenceAnnotation = (
   };
 };
 
+/**
+ * Check if the entity provides a `backstage.io/techdocs-ref` annotation of type `dir`
+ * and return the annotation. It accepts the two variants `<target>` and `dir:<target>`.
+ *
+ * @param entity the entity to check
+ */
+export const getDirLocation = (
+  entity: Entity,
+): { type: 'dir'; target: string } | undefined => {
+  const annotation = entity.metadata.annotations?.['backstage.io/techdocs-ref'];
+
+  if (!annotation) {
+    return undefined;
+  }
+
+  if (typeof annotation !== 'string') {
+    return undefined;
+  }
+
+  // if the string doesn't contain `:`, interpret it as the target of a dir type
+  if (!annotation.includes(':')) {
+    return { type: 'dir', target: annotation };
+  }
+
+  // note that `backstage.io/techdocs-ref: https://...` is invalid and will throw
+  const reference = parseLocationReference(annotation);
+
+  if (reference.type === 'dir') {
+    return {
+      type: 'dir',
+      target: reference.target,
+    };
+  }
+
+  // ignore any other types
+  return undefined;
+};
+
+/**
+ * TechDocs references of type `dir` are relative the source location of the entity.
+ * This function transforms relative references to absolute ones, based on the
+ * location the entity was ingested from. If the entity was registered by a `url`
+ * location, it returns a `url` location with a resolved target that points to the
+ * targeted subfolder. If the entity was registered by a `file` location, it returns
+ * an absolute `dir` location.
+ *
+ * @param entity the entity with annotations
+ * @param scmIntegrations access to the scmIntegrationt to do url transformations
+ * @throws if the entity doesn't specify a `dir` location or is ingested from an unsupported location.
+ * @returns the transformed location with an absolute target.
+ */
+export const transformDirLocation = (
+  entity: Entity,
+  scmIntegrations: ScmIntegrationRegistry,
+): { type: 'dir' | 'url'; target: string } => {
+  const dirLocation = getDirLocation(entity);
+
+  if (!dirLocation) {
+    throw new InputError(
+      `No techdocs location annotation provided in entity: ${stringifyEntityRef(
+        entity,
+      )}`,
+    );
+  }
+
+  const location = getEntitySourceLocation(entity);
+
+  switch (location.type) {
+    case 'url': {
+      const target = scmIntegrations.resolveUrl({
+        url: dirLocation.target,
+        base: location.target,
+      });
+
+      return {
+        type: 'url',
+        target,
+      };
+    }
+
+    case 'file': {
+      // only permit targets in the same folder as the target of the `file` location!
+      const target = resolveSafeChildPath(
+        path.dirname(location.target),
+        dirLocation.target,
+      );
+
+      return {
+        type: 'dir',
+        target,
+      };
+    }
+
+    default:
+      throw new InputError(`Unable to resolve location type ${location.type}`);
+  }
+};
+
 export const getLocationForEntity = (
```

Comment:
This function is also used to populate the location information to the `techdocs/metadata/entity/...` entpoint. This means the change is transparent to the UI as it only sees the resolved location.